### PR TITLE
WIP: NetStandard 2.0

### PR DESCRIPTION
**Pull request by tmat**
Retargets Roslyn projects to `netstandard2.0` and `net472`.

Moves most code from Workspaces.Desktop to Workspaces, except for MEF V1 implementation. 
Removes dependencies on Workspaces.Desktop.
Removes Workspaces.Desktop from Micrososft.CodeAnalysis.Workspaces.Common package.

Removes PortableFacades.vsix.

Enables more scripting tests on Core.

Fixes https://github.com/dotnet/roslyn/issues/29292 (Cleanup Microsoft.CodeAnalysis.Workspaces packages)
Fixes https://github.com/dotnet/roslyn/issues/18479 (Scripting Tuples test has broken references between desktop/coreclr tests)

To be merged only after VS validation.

### CODE REVIEW

**Comment by tmat** [comment]
test this please

**Comment by tmat** [comment]
test windows_debug_spanish_unit32_prtest please

**Comment by tmat** [comment]
test windows_debug_vs-integration_prtest  please

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest  please

**Comment by tmat** [comment]
test windows_release_unit32_prtest  please

**Comment by tmat** [comment]
@jaredpar @jasonmalinowski @dotnet/roslyn-infrastructure PTAL

**Comment by tmat** [comment]
@DustinCampbell Please take a look at the Workspaces related changes.

**Comment by tmat** [comment]
@jinujoe FYI

**Comment by tmat** [comment]
test windows_debug_vs-integration_prtest please

**Comment by tmat** [comment]
test windows_release_vs-integration_prtest please

File: build/Targets/BeforeCommonTargets.targets [hunk 1]
```diff
@@ -103,17 +103,19 @@
   </Choose>
 
   <!--
-    If TargetNetFX20 is true the project targets Framework 2.0 reference assemblies provided by Microsoft.NetFX20 nuget package.
+    If TargetNetFX20 is true the project targets Framework 2.0 reference assemblies provided by Microsoft.NetFramework.ReferenceAssemblies nuget package.
```
Comment (jasonmalinowski) [inline build/Targets/BeforeCommonTargets.targets]: "TargetNetFX20" still mentioned in the comment...

File: build/Targets/GenerateCompilerExecutableBindingRedirects.targets [hunk 2]
```diff
@@ -30,45 +30,9 @@
...
-      <SuggestedBindingRedirects Include="System.Threading.Thread, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
-        <MaxVersion>4.0.1.0</MaxVersion>
-      </SuggestedBindingRedirects>
-      <SuggestedBindingRedirects Include="System.Security.Cryptography.Primitives, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
-        <MaxVersion>4.0.1.0</MaxVersion>
-      </SuggestedBindingRedirects>
-      <SuggestedBindingRedirects Include="System.Security.Principal.Windows, Version=0.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a">
-        <MaxVersion>4.0.1.0</MaxVersion>
-      </SuggestedBindingRedirects>
-      <SuggestedBindingRedirects Include="System.ValueTuple, Version=0.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51">
```
Comment (jasonmalinowski) [inline build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: Any netstandard1.3 analyzers that use ValueTuple don't need this?

(guessing no, but I've seen confusion in the past where people forget about dependencies of analyzers, just wanted to double check)

File: build/Targets/Packages.props [hunk 3]
```diff
@@ -163,83 +163,20 @@
...
-    <SystemSecurityCryptographyX509CertificatesVersion>4.3.0</SystemSecurityCryptographyX509CertificatesVersion>
-    <SystemSecurityPrincipalWindowsVersion>4.3.0</SystemSecurityPrincipalWindowsVersion>
-    <SystemTextEncodingVersion>4.3.0</SystemTextEncodingVersion>
     <SystemTextEncodingCodePagesVersion>4.3.0</SystemTextEncodingCodePagesVersion>
-    <SystemTextEncodingExtensionsVersion>4.3.0</SystemTextEncodingExtensionsVersion>
-    <SystemTextRegularExpressionsVersion>4.3.0</SystemTextRegularExpressionsVersion>
-    <SystemThreadingVersion>4.3.0</SystemThreadingVersion>
-    <SystemThreadingTasksVersion>4.3.0</SystemThreadingTasksVersion>
     <SystemThreadingTasksDataflowVersion>4.5.24</SystemThreadingTasksDataflowVersion>
+    <!-- We need SVT assembly version at least 4.0.3.0 on net47 to make F5 work against Dev15 - see https://github.com/dotnet/roslyn/issues/29705 -->
```
Comment (jasonmalinowski) [inline build/Targets/Packages.props]: Nitpick: don't abbreviate assembly names in comments; it makes it harder to grep for stuff later.

File: build/scripts/test-build-correctness.ps1 [hunk 4]
```diff
@@ -31,7 +31,9 @@ try {
     # Verify the state of our various build artifacts
     Write-Host "Running BuildBoss"
     $buildBossPath = Join-Path $configDir "Exes\BuildBoss\BuildBoss.exe"
-    Exec-Console $buildBossPath "Roslyn.sln Compilers.sln SourceBuild.sln -r $repoDir $releaseArg"
+    $buildBossArgs = "Roslyn.sln Compilers.sln SourceBuild.sln -r $repoDir $releaseArg"
+    Write-Host "$buildBossPath $buildBossArgs"
+    Exec-Console $buildBossPath $buildBossArgs
```
Comment (jaredpar) [inline build/scripts/test-build-correctness.ps1]: Let me take an action item to add an option here to print the command line to the console. Should make this simpler for future PRs

File: build/Targets/Roslyn.Toolsets.Xunit.targets [hunk 5]
```diff
@@ -6,6 +6,35 @@
     <MSBuildAllProjects>$(MSBuildAllProjects);$(MSBuildThisFileFullPath)</MSBuildAllProjects>
   </PropertyGroup>
 
+  <ItemGroup>
+    <PackageReference Include="xunit" Version="$(XUnitVersion)" IsImplicitlyDefined="true" PrivateAssets="all"/>
+    <PackageReference Include="xunit.analyzers" Version="$(xunitanalyzersVersion)" IsImplicitlyDefined="true" PrivateAssets="all"/>
+    <PackageReference Include="xunit.runner.visualstudio" Version="$(XUnitRunnerVisualStudioVersion)" IsImplicitlyDefined="true" PrivateAssets="all"/>
+    <PackageReference Include="xunit.runner.console" Version="$(XUnitRunnerConsoleVersion)" IsImplicitlyDefined="true" PrivateAssets="all"/>
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="$(MicrosoftNETTestSdkVersion)" IsImplicitlyDefined="true"/>
+    <PackageReference Include="System.IO.FileSystem.Primitives" Version="$(SystemIOFileSystemPrimitivesVersion)" IsImplicitlyDefined="true"/>
```
Comment (jaredpar) [inline build/Targets/Roslyn.Toolsets.Xunit.targets]: Why do we need this reference?

File: build/Targets/Packages.props [hunk 6]
```diff
@@ -163,83 +163,20 @@
     <RoslynToolsVSIXExpInstallerVersion>1.0.0-beta2-63222-01</RoslynToolsVSIXExpInstallerVersion>
     <RoslynToolsMSBuildVersion>0.5.0-alpha</RoslynToolsMSBuildVersion>
     <RoslynToolsNuGetRepackVersion>1.0.0-beta2-63223-01</RoslynToolsNuGetRepackVersion>
-    <RoslynToolsReferenceAssembliesVersion>0.1.3</RoslynToolsReferenceAssembliesVersion>
```
Comment (jaredpar) [inline build/Targets/Packages.props]: Yet another of my hacks gone.

File: src/Compilers/VisualBasic/Portable/Compilation/ClsComplianceChecker.vb [hunk 7]
```diff
@@ -777,10 +777,10 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         Private Function GetDeclaredComplianceHelper(symbol As Symbol, <Out> ByRef attributeLocation As Location, <Out> ByRef isAttributeInherited As Boolean) As Boolean?
             attributeLocation = Nothing
             isAttributeInherited = False
-            For Each data In symbol.GetAttributes()
+            For Each attributeData In symbol.GetAttributes()
```
Comment (jaredpar) [inline src/Compilers/VisualBasic/Portable/Compilation/ClsComplianceChecker.vb]: Was this change necessary?

File: src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.swr [hunk 8]
```diff
@@ -10,62 +10,36 @@ vs.nonCriticalProcesses
...
-    file source=$(OutputPath)\Vsix\Roslyn.Compilers.Extension\System.Xml.ReaderWriter.dll vs.file.ngenArchitecture=all vs.file.ngenApplication="[installDir]\MSBuild\15.0\Bin\Roslyn\VBCSCompiler.exe"
-    file source=$(OutputPath)\Vsix\Roslyn.Compilers.Extension\System.Xml.XmlDocument.dll vs.file.ngenArchitecture=all vs.file.ngenApplication="[installDir]\MSBuild\15.0\Bin\Roslyn\VBCSCompiler.exe"
-    file source=$(OutputPath)\Vsix\Roslyn.Compilers.Extension\System.Xml.XPath.dll vs.file.ngenArchitecture=all vs.file.ngenApplication="[installDir]\MSBuild\15.0\Bin\Roslyn\VBCSCompiler.exe"
-    file source=$(OutputPath)\Vsix\Roslyn.Compilers.Extension\System.Xml.XPath.XDocument.dll vs.file.ngenArchitecture=all vs.file.ngenApplication="[installDir]\MSBuild\15.0\Bin\Roslyn\VBCSCompiler.exe"
-
-# We never load these binaries, and it does not generate the correct binding redirects, so do not NGEN them
-# https://github.com/dotnet/roslyn/pull/27537
-    file source=$(OutputPath)\Vsix\Roslyn.Compilers.Extension\System.Net.Http.dll
-    file source=$(OutputPath)\Vsix\Roslyn.Compilers.Extension\System.Diagnostics.DiagnosticSource.dll
```
Comment (jaredpar) [inline src/Setup/DevDivVsix/CompilersPackage/Microsoft.CodeAnalysis.Compilers.swr]: Glorious red.

File: src/Tools/BuildBoss/ProjectCheckerUtil.cs [hunk 9]
```diff
@@ -441,8 +441,7 @@ private bool CheckTargetFrameworks(TextWriter textWriter, RoslynProjectData data
                 switch (targetFramework)
                 {
                     case "net20":
-                    case "net46":
-                    case "netstandard1.3":
+                    case "net472":
                     case "netcoreapp1.1":
```
Comment (jaredpar) [inline src/Tools/BuildBoss/ProjectCheckerUtil.cs]: Is this valid anymore?

**Comment by jaredpar** [comment]
Done with review pass (iteration 21)

File: build/Targets/BeforeCommonTargets.targets (hunk 1 above)
Comment (tmat) [reply build/Targets/BeforeCommonTargets.targets]: Fixed.

File: build/Targets/GenerateCompilerExecutableBindingRedirects.targets (hunk 2 above)
Comment (tmat) [reply build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: Good question.

File: build/Targets/Packages.props (hunk 3 above)
Comment (tmat) [reply build/Targets/Packages.props]: Fixed.

File: src/VisualStudio/Setup.Dependencies/Roslyn.VisualStudio.Setup.Dependencies.csproj [hunk 10]
```diff
@@ -76,51 +46,36 @@
...
+    The binding redirects are specified manually since ProvideDependencyBindingRedirection attribute doesn't generate the right version for some reason.
+    Using System.ValueTuple.dll from Framework directory does not work since we'd need a downgrading redirect 0.0.0.0-4.0.1.0 -> 4.0.0.0, which VS doesn't apply
+    since it already has a redirect 0.0.0.0-4.0.1.0 -> 4.0.1.0.
+    
+    TODO: Remove once we move to Dev16 (https://github.com/dotnet/roslyn/issues/29705)
+  -->
+  <ItemGroup>
+    <PackageReference Include="System.ValueTuple" Version="$(SystemValueTupleVersion)" />
+    <NuGetPackageToIncludeInVsix Include="System.ValueTuple" />
+    <Content Include="Roslyn.VisualStudio.Setup.Dependencies.ValueTuple.pkgdef" IncludeInVSIX="true" />
```
[the remark this answers is not in the document]
Comment (jasonmalinowski) [reply src/VisualStudio/Setup.Dependencies/Roslyn.VisualStudio.Setup.Dependencies.csproj]: But peeking anyways, why a manual .pkgdef vs. just putting this inside AssemblyRedirects like everything else?

File: src/VisualStudio/Setup.Dependencies/AssemblyRedirects.cs [hunk 11]
```diff
@@ -1,37 +1,11 @@
...
-[assembly: ProvideDependencyBindingRedirection("System.Security.Cryptography.Encoding.dll")]
-[assembly: ProvideDependencyBindingRedirection("System.Security.Cryptography.Primitives.dll")]
-[assembly: ProvideDependencyBindingRedirection("System.Security.Cryptography.X509Certificates.dll")]
-//[assembly: ProvideDependencyBindingRedirection("System.Text.Encoding.dll")] // no implementation assembly for net46
-//[assembly: ProvideDependencyBindingRedirection("System.Text.Encoding.Extensions.dll")] // no implementation assembly for net46
-//[assembly: ProvideDependencyBindingRedirection("System.Text.RegularExpressions.dll")] // no implementation assembly for net46
 [assembly: ProvideDependencyBindingRedirection("System.Threading.Tasks.Extensions.dll")]
-//[assembly: ProvideDependencyBindingRedirection("System.Threading.Tasks.Parallel.dll")] // no implementation assembly for net46
 [assembly: ProvideDependencyBindingRedirection("System.Text.Encoding.CodePages.dll")]
-[assembly: ProvideDependencyBindingRedirection("System.ValueTuple.dll")]
```
Comment (jasonmalinowski) [inline src/VisualStudio/Setup.Dependencies/AssemblyRedirects.cs]: Commented elsewhere too: not sure why you're removing this and re-adding a pkgdef doing it manually. I'm guessing it's either something subtly different with the redirect produced or "that way the hack is all in one place". If the latter, it seems we're just confusing the situation by this file actually "missing" something.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs [hunk 12]
```diff
@@ -101,9 +101,17 @@ private bool VsSmartScopeCandidate(string fullPath)
             return _runtimeDirectories.Any(d => fullPath.StartsWith(d, StringComparison.OrdinalIgnoreCase));
         }
 
+        internal static IEnumerable<string> GetReferencePaths()
+        {
+            // TODO:
+            // WORKAROUND: properly enumerate them
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/MetadataReferences/VisualStudioMetadataReferenceManager.cs]: Is this TODO new?

File: src/VisualStudio/Core/Def/Implementation/GenerateType/GenerateTypeDialogViewModel.cs [hunk 13]
```diff
@@ -319,7 +299,11 @@ internal bool TrySubmit()
                 if (lastIndexOfSeparatorInFullPath != -1)
                 {
                     var fileNameInFullPathInContainers = this.FullFilePath.Split(new[] { '\\' }, StringSplitOptions.RemoveEmptyEntries);
-                    this.FullFilePath = string.Join("\\", fileNameInFullPathInContainers.Select(str => str.TrimStart()));
+
+                    // Trim spaces of each component of the file name.
+                    // Note that path normalization changed between 4.6.1 and 4.6.2 and GetFullPath no longer trims trailing spaces.
```
Comment (jasonmalinowski) [inline src/VisualStudio/Core/Def/Implementation/GenerateType/GenerateTypeDialogViewModel.cs]: So we're already broken on machines that were upgraded? Do we not have tests for this?

File: src/Test/Utilities/Portable/FX/EnsureInvariantCulture.cs [hunk 14]
```diff
@@ -17,11 +17,7 @@ public EnsureInvariantCulture()
             _threadId = Thread.CurrentThread.ManagedThreadId;
             _threadCulture = CultureInfo.CurrentCulture;
 
-#if DNX
```
Comment (jasonmalinowski) [inline src/Test/Utilities/Portable/FX/EnsureInvariantCulture.cs]: Hah!

File: src/Setup/DevDivPackages/Debugger/DevDivPackagesDebugger_DoNotUse.csproj [hunk 15]
```diff
@@ -22,15 +22,5 @@
   </ItemGroup>
   <ItemGroup>
     <PackageReference Include="Microsoft.DiaSymReader.PortablePdb" Version="$(MicrosoftDiaSymReaderPortablePdbVersion)" />
```
Comment (jasonmalinowski) [inline src/Setup/DevDivPackages/Debugger/DevDivPackagesDebugger_DoNotUse.csproj]: Does this file really need to exist now, given it's referencing just a single thing we also reference?

File: src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj [hunk 16]
```diff
@@ -102,6 +102,8 @@
   <ItemGroup>
     <PackageReference Include="Microsoft.VisualStudio.SDK.Analyzers" Version="$(MicrosoftVisualStudioSDKAnalyzersVersion)" PrivateAssets="all" />
     <PackageReference Include="Microsoft.VisualStudio.Shell.15.0" Version="$(MicrosoftVisualStudioShell150Version)" />
+    <PackageReference Include="Microsoft.VisualStudio.Threading" Version="$(MicrosoftVisualStudioThreadingVersion)" />
+    <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonVersion)" />
```
Comment (jasonmalinowski) [inline src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj]: Why? This bit seems surprising.

File: build/Targets/Roslyn.Toolsets.Xunit.targets (hunk 5 above)
Comment (jasonmalinowski) [reply build/Targets/Roslyn.Toolsets.Xunit.targets]: What is tracking?

File: src/VisualStudio/Setup.Dependencies/Roslyn.VisualStudio.Setup.Dependencies.csproj [hunk 17]
```diff
@@ -30,40 +30,10 @@
     -->
     <NuGetPackageToIncludeInVsix Include="Microsoft.DiaSymReader" />
     <NuGetPackageToIncludeInVsix Include="Microsoft.DiaSymReader.PortablePdb" />
-    <NuGetPackageToIncludeInVsix Include="System.AppContext" />
```
[the remark this answers is not in the document]
Comment (tmat) [reply src/VisualStudio/Setup.Dependencies/Roslyn.VisualStudio.Setup.Dependencies.csproj]: It works.

File: src/VisualStudio/Core/Def/Implementation/GenerateType/GenerateTypeDialogViewModel.cs (hunk 13 above)
Comment (tmat) [reply src/VisualStudio/Core/Def/Implementation/GenerateType/GenerateTypeDialogViewModel.cs]: We have been broken if our dll was loaded to an .exe that was built against net462+. We are now using xunit runner built against net472, so that's where it showed up.

File: src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj (hunk 16 above)
Comment (tmat) [reply src/ExpressionEvaluator/Package/ExpressionEvaluatorPackage.csproj]: I don't know. NuGet magic collided with RAR magic.

File: src/Setup/DevDivPackages/Debugger/DevDivPackagesDebugger_DoNotUse.csproj (hunk 15 above)
Comment (tmat) [reply src/Setup/DevDivPackages/Debugger/DevDivPackagesDebugger_DoNotUse.csproj]: Nope, will remove it

File: build/Targets/GenerateCompilerExecutableBindingRedirects.targets (hunk 2 above)
Comment (tmat) [reply build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: Yes, but interestingly enough turns out that we are now missing the following redirects:

```
  <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Reflection.TypeExtensions" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.1.1.0" newVersion="4.1.1.0" />
      </dependentAssembly>
    </assemblyBinding>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Security.AccessControl" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
      </dependentAssembly>
    </assemblyBinding>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Security.Principal.Windows" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.0.1.0" newVersion="4.0.1.0" />
      </dependentAssembly>
    </assemblyBinding>
    <assemblyBinding xmlns="urn:schemas-microsoft-com:asm.v1">
      <dependentAssembly>
        <assemblyIdentity name="System.Text.Encoding.CodePages" publicKeyToken="b03f5f7f11d50a3a" culture="neutral" />
        <bindingRedirect oldVersion="0.0.0.0-4.1.1.0" newVersion="4.1.1.0" />
      </dependentAssembly>
    </assemblyBinding>
```

We have been missing a couple previously. A couple are new. 

I created an app that targets net46 and net472 and depends on a library that targets netstandard1.3. The library uses types from all the portable facades that we have currently in VS. And the binding redirects above were generated by msbuild.

Comment (tmat) [reply build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: Actually, we don't ship `System.Reflection.TypeExtensions`, `System.Security.AccessControl` and `System.Security.Principal.Windows` with the compiler, so analyzers can't use them anyways, unless they carry the assemblies with them, in which case they should have the right versions and not need binding redirects.

Comment (tmat) [reply build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: To be certain I have added tests that validates we won't break analyzers built against netstandard1.3 and using all sorts of APIs that are available there. https://github.com/dotnet/roslyn/pull/29805

Comment (tmat) [reply build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: The test passes with this PR as is.

Comment (tmat) [reply build/Targets/GenerateCompilerExecutableBindingRedirects.targets]: The redirects I mentioned above are not actually needed. I haven't emulated Roslyn correctly in that test app. E.g. the App referenced System.Text.Encoding.CodePages 4.5.0, but Roslyn uses 4.3.0. These result in different assembly version actually used and thus different required redirects.

File: src/Compilers/VisualBasic/Portable/Compilation/ClsComplianceChecker.vb (hunk 7 above)
Comment (jaredpar) [reply src/Compilers/VisualBasic/Portable/Compilation/ClsComplianceChecker.vb]: Had not thought about that. @KathleenDollard you may be interested in that.

**Comment by tmat** [comment]
Superseded by https://github.com/dotnet/roslyn/pull/30914
